torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -912,3 +923,182 @@ def save_checkpoint(
                     "The full model checkpoint, including all weights and configurations, has been saved successfully."
                     "You can now use this checkpoint for further training or inference."
                 )
+
+
+class DistributedCheckpointer(_CheckpointerInterface):
+    """
+    Checkpointer which reads and writes checkpoints in the DistributedCheckpointing format.
+
+    Args:
+        checkpoint_dir (str): Directory containing the checkpoint files
+        output_dir (str): Directory to save the checkpoint files
+    """
+
+    def __init__(
+        self,
+        checkpoint_dir: str,
+        output_dir: str,
+    ) -> None:
+        self._checkpoint_dir = Path(checkpoint_dir)
+        self._output_dir = Path(output_dir)
+        self._checkpoint_future = None
+        self._checkpoint_dir_prefix = "checkpoint"
+        self._metadata_file = ".metadata"
+        _, rank = training.get_world_size_and_rank()
+        self._is_rank_zero = rank == 0
+
+    def _get_latest_intermediate_checkpoint_path(self) -> Optional[str]:
+        """
+        This method iterates over the available intermediate checkpoints and finds the latest checkpoint
+        to load.
+
+        Returns:
+            str: The fully qualified path of the checkpoint directory containing the latest and valid
+            intermediate checkpoint
+        """
+
+        checkpoint_dir_pattern = re.compile(f"{self._checkpoint_dir_prefix}_(\\d+)")
+        checkpoint_paths = [
+            name
+            for name in os.listdir(self._output_dir)
+            if re.match(checkpoint_dir_pattern, name)
+            and os.path.isfile(
+                os.path.join(self._output_dir, name, self._metadata_file)
+            )
+        ]
+
+        if checkpoint_paths:
+            latest_checkpoint_dir = sorted(checkpoint_paths)[-1]
+            return os.path.join(self._output_dir, latest_checkpoint_dir)
+        return None
+
+    def load_checkpoint(self, state_dict: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Load a Distributed checkpoint
+        """
+        checkpoint_path = self._get_latest_intermediate_checkpoint_path()
+
+        if checkpoint_path is None:
+            raise ValueError(
+                "No intermediate checkpoint found in the specified directory. Please ensure that a checkpoint exists."
+            )
+
+        if self._is_rank_zero:
```

Comment:
you can use the `torchtune.utils.logging.log_rank_zero` utility so you don't have to keep checking for rank

Reply:
Yeah. DCP APIs are trainer agnostic. So really all the code in the integration is gonna be state dict init & updates along with the tests.